host/src/operations/mod.rs

```diff
@@ -0,0 +1,329 @@
+        } else {
+            panic!(
+                "Workflow exited: {} - | err: {}",
+                res.status,
+                res.error_msg.unwrap_or_default()
+            );
+        }
+    }
+}
+
+pub async fn maybe_prove<I: Serialize, O: Eq + Debug + Serialize + DeserializeOwned>(
+    cli: &Cli,
+    input: &I,
+    elf: &[u8],
+    expected_output: &O,
+    assumptions: (Vec<Assumption>, Vec<String>),
+) -> Option<(String, Receipt)> {
+    let Cli::Prove(prove_args) = cli else {
+        return None;
+    };
+
+    let (assumption_instances, assumption_uuids) = assumptions;
+    let encoded_input = to_vec(input).expect("Could not serialize proving input!");
+
+    let encoded_output =
+        to_vec(expected_output).expect("Could not serialize expected proving output!");
+    let computed_image_id = compute_image_id(elf).expect("Failed to compute elf image id!");
+
+    let receipt_label = format!(
+        "{}-{}",
+        hex::encode(computed_image_id),
+        hex::encode(keccak(bytemuck::cast_slice(&encoded_output)))
+    );
+
+    // get receipt
+    let (mut receipt_uuid, receipt, cached) =
+        if let Ok(Some(cached_data)) = load_receipt(&receipt_label) {
+            info!("Loaded locally cached receipt");
+            (cached_data.0, cached_data.1, true)
+        } else if prove_args.submit_to_bonsai {
+            // query bonsai service until it works
+            loop {
+                if let Ok(remote_proof) = prove_bonsai(
+                    encoded_input.clone(),
+                    elf,
+                    expected_output,
+                    assumption_uuids.clone(),
+                )
+                .await
+                {
+                    break (remote_proof.0, remote_proof.1, false);
+                }
+            }
+        } else {
+            // run prover
+            (
+                Default::default(),
+                prove_locally(
+                    prove_args.exec_args.execution_po2,
+                    encoded_input,
+                    elf,
+                    assumption_instances,
+                    prove_args.exec_args.profile,
+                    &cli.execution_label(),
+                ),
+                false,
+            )
+        };
+
+    // verify output
+    let output_guest: O = receipt.journal.decode().unwrap();
+    if expected_output == &output_guest {
+        info!("Prover succeeded");
+    } else {
+        error!(
+            "Output mismatch! Prover: {:?}, expected: {:?}",
+            output_guest, expected_output,
+        );
+    }
+
+    // upload receipt to bonsai
+    if prove_args.submit_to_bonsai && receipt_uuid.is_empty() {
+        info!("Uploading cached receipt without UUID to Bonsai.");
+        receipt_uuid = upload_receipt(&receipt)
+            .await
+            .expect("Failed to upload cached receipt to Bonsai");
+    }
+
+    let result = (receipt_uuid, receipt);
+
+    // save receipt
+    if !cached {
```

Comment:
The save happens if the receipt was not loaded from the cache, as to just avoid rewriting the same data that was just read.

It should indeed respect the argument I guess to avoid confusingly writing data when it may imply that it shouldn't.